src/sidebar.c

```diff
@@ -418,45 +429,319 @@ static gchar *get_doc_folder(const gchar *path)
+	enum {
+		TREE_CASE_NONE,
+		TREE_CASE_EQUALS,
+		TREE_CASE_CHLID_OF,
+		TREE_CASE_PARENT_OF,
+		TREE_CASE_HAVE_SAME_PARENT
+	} best_case;
+} TreeForeachData;
+
+
+static gboolean tree_foreach_callback(GtkTreeModel *model, GtkTreePath *path,
+					GtkTreeIter *iter, gpointer data_in)
+{
+	gchar *name;
+	gchar *dirname;
+	guint diff;
+	gsize name_len;
+	GeanyDocument *doc;
+	TreeForeachData *data = (TreeForeachData*)data_in;
+
+	gtk_tree_model_get(model, iter,
+			DOCUMENTS_FILENAME, &name,
+			DOCUMENTS_DOCUMENT, &doc, -1);
+
+	if (doc) /* skip documents */
+		goto finally;
+
+	dirname = get_doc_folder(name);
+	if (dirname)
+		SETPTR(name, dirname);
+
+	diff = pathcmp(name, data->needle);
+	name_len = strlen(name);
+
+	if (diff == 0)
+		goto finally;
+
+	if (data->best_len < diff)
+	{
+		gint best_case;
+		gboolean tree = documents_show_paths == SHOW_PATHS_TREE;
+
+		/* there are four cases */
+		/* first case: exact match. File is from already opened dir */
+		if (name_len == diff && data->needle_len == name_len)
+			best_case = TREE_CASE_EQUALS;
+		/* second case: split current dir. File is from deeper level */
+		else if (name_len == diff && tree)
+			best_case = TREE_CASE_CHLID_OF;
+		/* third case: split parent dir. File is from one of existing level */
+		else if (data->needle_len == diff && tree)
+			best_case = TREE_CASE_PARENT_OF;
+		/* fourth case: both dirs have same parent */
+		else if (tree)
+			best_case = TREE_CASE_HAVE_SAME_PARENT;
+		else
+			goto finally;
+		data->best_len = diff;
+		data->best_case = best_case;
+		data->best_iter = *iter;
+	}
+finally:
+	g_free(name);
+	return FALSE;
+}
+
+/* Returns TRUE if parent points to a newly added row,
+ * caller might want to expand the relevant rows in the tree view */
+static gboolean get_doc_parent(GeanyDocument *doc, GtkTreeIter *parent)
+{
+	gchar *path;
+	GtkTreeIter iter;
+	gint name_diff = 0;
+	gboolean has_parent;
+	GtkTreeModel *model = GTK_TREE_MODEL(store_openfiles);
+	TreeForeachData data = {NULL, 0, 0, {0}, TREE_CASE_NONE};
+
+	path = g_path_get_dirname(DOC_FILENAME(doc));
+
+	/* find best opened dir */
+	data.needle = get_doc_folder(path);
+	data.needle_len = strlen(data.needle);
+	name_diff = strlen(path) - data.needle_len;
+	gtk_tree_model_foreach(model, tree_foreach_callback, (gpointer)&data);
+
+	switch (data.best_case)
+	{
+		case TREE_CASE_EQUALS:
+		{
+			*parent = data.best_iter;
+			break;
+		}
+		case TREE_CASE_CHLID_OF:
+		{
+			/* This dir is longer than existing so just add child */
+			tree_add_new_dir(parent, &data.best_iter, path);
+			break;
+		}
+		case TREE_CASE_PARENT_OF:
+		{
+			/* More complicated logic. This dir should be a parent
+			 * of existing, so reparent existing dir.  */
+			has_parent = gtk_tree_model_iter_parent(model, &iter, &data.best_iter);
+			tree_add_new_dir(parent, has_parent ? &iter : NULL, path);
+			tree_copy_recursive(&data.best_iter, parent);
+			gtk_tree_store_remove(store_openfiles, &data.best_iter);
```

Comment:
Looking at how it's used, it could be nice to make a `tree_reparent()` or something, which would do this plus removing the old parent.  That the implementation requires copying is sad, but would be encapsulated and wouldn't get people like me to look suspiciously at the calling code before going further :)

Reply:
Looking at this, your suggestion sounds good and would also make the intent of the function more clear.

The copy is currently mainly done because the shortname of a directory node may change (if a new parent is added in-between then child nodes could change from a/b/c/d to a->b/c/d, for example (after opening a/x.c))

Reply:
Oh, and because gtk_tree_store_move_after/_before don't work on sorted tree stores.